Add follow notifications for soapbox accounts

MrPetovan opened the pull request:
Closes #4639

annando commented on static/dbstructure.config.php:
```diff
@@ -932,6 +933,7 @@
 			"uid_type_link" => ["uid", "type", "link(190)"],
 			"uri-id" => ["uri-id"],
 			"parent-uri-id" => ["parent-uri-id"],
+			"uid_cid" => ["uid", "cid", "type"],
```
Is `type` really needed here? if yes, then please add it to the name of the index as well.

MrPetovan replied:
Yeah, to filter by type in `Notify->deleteForUserByContactId`

annando commented on src/Model/Contact.php:
```diff
@@ -2737,17 +2743,36 @@ public static function addRelationship(array $importer, array $contact, array $d
 	public static function removeFollower(array $contact)
 	{
+		// Remove follow request notifications
+		DI::notify()->deleteForUserByContactId($contact['uid'], $contact['id'], [Type::INTRO]);
```
Please ensure to apply the changes also to the `notification`  table. We really should drop this old `notify` table and finally do everything with the `notification` table.

annando commented on src/Model/Contact.php:
```diff
@@ -2663,6 +2660,15 @@ public static function addRelationship(array $importer, array $contact, array $d
 
 			Post\UserNotification::insertNotification($contact['id'], Activity::FOLLOW, $importer['uid']);
 
+			DI::notify()->createFromArray([
```
The `notification` table eventually should completely replace the `notify` table, also the `notify` table is updated via the function above which then calls this function: https://github.com/friendica/friendica/blob/develop/src/Model/Subscription.php#L157

annando commented on mod/ping.php:
```diff
@@ -401,12 +401,15 @@ function ping_get_notifications($uid)
+			ORDER BY `notify`.`date` $order
+			LIMIT ?, 50",
 			$uid,
-			Notification\Type::INTRO,
```
Doesn't this change adds such notifications for all users?

MrPetovan replied:
Yes, this is the "bad" part of this. Since we fetch the introductions independently, these notifications have been filtered out, but in the case of soapbox/freelove accounts, there's no introduction to match on. I guess I should make it a dynamic array depending on the user's account type.

annando reviewed (changes requested):
We have all the needed data in the `notification` table. Since we really should drop the `notify` table, I prefer a solution where the `ping.php` just fetches the data from the `nofitication` table instead, without doing anything with the `notify` table.

MrPetovan commented:
I can never remember which one from `notification` or `notify` needs to go, thanks for the reminder.

annando commented:
Yeah. How can we document which table to favour in the future?

MrPetovan commented:
Add a "Deprecated" in the `notify` table comment and in all the classes related to the table.

MrPetovan committed:
Deprecated the notify table/classes

MrPetovan committed:
Use public contact ID in Model\Post\UserNotification::insertNotification

- Add localRelationship dependency to Notification factory
- Remove dependencies from Factory\Notification->getMessageFromNotification method

MrPetovan committed:
Add legacy notification endpoint for deprecated Notify entity

- Add dependencies to Notifications\Notification module class

MrPetovan committed:
Replace Javascript notification string formatting with Smarty templates

MrPetovan committed:
Remove unused deprecated XML output in mod/ping

- Remove unused parameters localUser and uid in mod/ping
- Unify data array initialization in mod/ping

MrPetovan committed:
Move mod/ping to module class

MrPetovan committed:
Switch data source from notify to notification in Module\Notifications\Ping

- Change unused FormattedNotification classes to FormattedNavNotification classes

MrPetovan committed:
Delete follow notification on unfollow

- Fix translation string typo

MrPetovan committed:
Updated main translation file after string changes

MrPetovan commented:
This PR is now complete but I'm waiting on version 2020.03 to be released.

We are now using the `notification` table for the notification menu, but I didn't check all the places we are still only using the `notify` table.

I also wasn't able to test introductions (follow requests) on my dev node, when I switch my account type to `Standard` remote contacts still are able to follow me directly. I assume it is a quirk limited to my dev node otherwise we would have heard about it.

annando commented:
This looks ... intense :grimacing:. Can you run it for some time on your production system? This is too huge to check everything just by looking at the code.

MrPetovan commented:
For reviewing this PR, I suggest to go commit by commit to reduce the complexity. Beyond that, running this branch just on my own system wouldn't be meaningful because I simply wouldn't know about the notifications I would potentially miss, so I'd prefer if we took some time to test it together or with someone else.

It's already running on my dev node and notifications are showing which is what this PR is about, so I'm confident I didn't make a programming mistake, and now we have to test the usability.

annando commented:
Are you looking at the conflict? I hadn't had a deep look at the changes now. Since `notify` is now deprecated, how are you solving the systems notifications that aren't covered by the `notification` table?

MrPetovan commented:
Deprecated doesn't mean removed. It just means that we won't expand it. Current uses of `notify` include the notification module that needs reworking, and email notifications that should be separated from `notify` handling. Everything else should be covered by `notification`.

Edit: Conflict fixed.

MrPetovan commented:
Thank you for the merge, I'll issue a notice on the Friendica Support forum to be on the lookup for notifications mishaps.

annando commented on src/Navigation/Notifications/Repository/Notification.php:
```diff
@@ -100,6 +100,36 @@ public function selectForUser(int $uid, array $condition = [], array $params = [
+	 */
+	public function selectDigestForUser(int $uid): Collection\Notifications
+	{
+		$rows = $this->db->p("
```
I haven't recognized this query when looking at the code. What is this query supposed to do, why is it so complicated?

MrPetovan replied:
This is the query that determines the most recent notification for any given thread. How else would you do it?

annando replied:
I don't see the need for a subquery to the same table. Also the `group by` values are looking weird.

MrPetovan replied:
Again, how else would you do it? Not all notifications are pertaining to a thread, so the next grouping mechanism is by actor to get the most recent notification from the same contact. This is the purpose of the `IFNULL()` call, as a fallback when `parent-uri-id` isn't set.

annando replied:
I have to make some tests to check which query is the best.

MrPetovan replied:
I use a ``IN (SELECT MAX(`id`) GROUP BY)`` because MySQL's behavior with the order of the lines matched by a GROUP BY clause isn't defined. This is a consequence of allowing non-GROUP BY columns in the SELECT. To consistently retrieve the most recent row per matched group, you need to perform an aggregate function (here, `MAX()`) to obtain the id of the expected row, and then match on this id to retrieve the complete record.

I do not know how to do it differently in (My)SQL.

annando replied:
I played around with that and other queries. This one seems to be okay. I guess it was my
i
can't
read
and
understand
queries
when they are split in a lot of lines-problem :-)

MrPetovan replied:
I find having a single keyword on each line to be more readable, but it isn't the first time you've exhibited programming chaos-loving behavior!
